grpc/src/client.rs

```diff
@@ -519,125 +554,177 @@ impl GrpcClient {
-        })
+
+            break res?;
+        };
+
+        Ok((usage, cfg.gas_price.unwrap_or(price)))
     }
 
     async fn sign_and_broadcast_blobs(
         &self,
         blobs: Vec<Blob>,
         cfg: TxConfig,
         context: &Context,
-    ) -> Result<(Hash, u64)> {
+    ) -> Result<BroadcastedTx> {
         let ChainState { chain_id, .. } = self.load_chain_state(context).await?;
         // lock the account; tx signing and broadcast must be atomic
         // because node requires all transactions to be sequenced by account.sequence
-        let (account, signer) = self.load_account(context).await?;
+        let mut account = self.lock_account(context).await?;
 
-        let pfb = MsgPayForBlobs::new(&blobs, account.address.clone())?;
+        let pfb = MsgPayForBlobs::new(&blobs, account.base.address)?;
         let pfb = RawTxBody {
             messages: vec![RawMsgPayForBlobs::from(pfb).into_any()],
             memo: cfg.memo.clone().unwrap_or_default(),
             ..RawTxBody::default()
         };
+        let blobs: Vec<_> = blobs.into_iter().map(Into::into).collect();
 
         let (gas_limit, gas_price) = self
-            .calculate_transaction_gas_params(&pfb, &cfg, chain_id.clone(), &account, context)
+            .calculate_transaction_gas_params(&pfb, &cfg, chain_id.clone(), &mut account, context)
             .await?;
-
         let fee = (gas_limit as f64 * gas_price).ceil() as u64;
-        let tx = sign_tx(
-            pfb,
-            chain_id.clone(),
-            &account,
-            &signer.pubkey,
-            &signer.signer,
-            gas_limit,
-            fee,
-        )
-        .await?;
-
-        let blobs = blobs.into_iter().map(Into::into).collect();
-        let blob_tx = RawBlobTx {
-            tx: tx.encode_to_vec(),
-            blobs,
-            type_id: BLOB_TX_TYPE_ID.to_string(),
-        };
 
-        self.broadcast_tx_with_account(blob_tx.encode_to_vec(), cfg, account, context)
-            .await
+        // in case parallel submission failed because of the sequence mismatch,
+        // we update the account, resign and broadcast transaction
+        loop {
+            let tx = sign_tx(
+                pfb.clone(),
+                chain_id.clone(),
+                &account.base,
+                account.pubkey,
+                account.signer,
+                gas_limit,
+                fee,
+            )
+            .await?;
+
+            let blob_tx = RawBlobTx {
+                tx: tx.encode_to_vec(),
+                blobs: blobs.clone(),
+                type_id: BLOB_TX_TYPE_ID.to_string(),
+            }
+            .encode_to_vec();
+
+            let res = self
+                .broadcast_tx_with_account(blob_tx, &cfg, &mut account, context)
+                .await;
+
+            if let Some(new_sequence) = res.as_ref().err().and_then(extract_sequence_on_mismatch) {
+                account.base.sequence = new_sequence?;
+                continue;
```

Comment:
This goes back to `sign_tx`, shouldn't we recalculate gas?

Reply:
So - in theory, we may hit this error and retry when new block is produced, meaning the gas prices could change. But the estimation service takes median from a few consecutive blocks, so the changes to gas price should be fairly amortized by that.

Celestia-app doesn't do it, it just [recursively calls the broadcast of transaction](https://github.com/celestiaorg/celestia-app/blob/main/pkg/user/tx_client.go#L550). Although in our code it would look quite nice to do this, we could remove the loop from `calculate_transaction_gas_params` and react to sequence errors only within the `broadcast_xyz` methods.
wdyt? imo it sounds good